Test/Microsoft.ApplicationInsights.Test/Shared/Extensibility/Implementation/Tracing/HeartbeatTests.cs

```diff
@@ -0,0 +1,357 @@
+        {
+            using (var hbeat = new HeartbeatProvider() { HeartbeatInterval = TimeSpan.FromMilliseconds(0) })
+            {
+                hbeat.Initialize(configuration: null);
+                Assert.AreEqual(hbeat.HeartbeatInterval, HeartbeatProvider.DefaultHeartbeatInterval);
+            }
+        }
+
+        [TestMethod]
+        public void CanExtendHeartbeatPayload()
+        {
+
+            using (var hbeat = new HeartbeatProvider())
+            {
+                hbeat.Initialize(configuration: new TelemetryConfiguration());
+
+                try
+                {
+                    Assert.IsTrue(hbeat.AddHealthProperty("test01", "this is a value", true));
+                }
+                catch (Exception e)
+                {
+                    Assert.Fail(string.Format(CultureInfo.CurrentCulture, "Registration of a heartbeat payload provider throws exception '{0}", e.ToInvariantString()));
+                }
+            }
+        }
+
+        [TestMethod]
+        public void CanSetDelayBetweenHeartbeats()
+        {
+            TimeSpan userSetInterval = TimeSpan.FromMilliseconds(7252.0);
+
+            using (var hbeat = new HeartbeatProvider())
+            {
+                var config = new TelemetryConfiguration(string.Empty, new StubTelemetryChannel());
+                hbeat.Initialize(configuration: config);
+                Assert.AreNotEqual(userSetInterval, hbeat.HeartbeatInterval.TotalMilliseconds);
+
+                hbeat.HeartbeatInterval = userSetInterval;
+                hbeat.Initialize(configuration: config);
+                Assert.AreEqual(userSetInterval, hbeat.HeartbeatInterval);
+            }
+        }
+
+        [TestMethod]
+        public void DiagnosticsTelemetryModuleCreatesHeartbeatModule()
+        {
+            using (var diagModule = new DiagnosticsTelemetryModule())
+            {
+                diagModule.Initialize(new TelemetryConfiguration());
+                Assert.IsNotNull(diagModule.HeartbeatProvider);
+            }
+        }
+
+        [TestMethod]
+        public void HeartbeatPayloadContainsDataByDefault()
+        {
+            using (var hbeat = new HeartbeatProvider())
+            {
+                var config = new TelemetryConfiguration(string.Empty, new StubTelemetryChannel());
+                hbeat.Initialize(configuration: config);
+                var hbeatPayloadData = hbeat.GatherData();
+                Assert.IsNotNull(hbeatPayloadData);
+            }
+        }
+
+        [TestMethod]
+        public void HeartbeatPayloadContainsUserSpecifiedData()
+        {
+            using (var hbeat = new HeartbeatProvider())
+            {
+                var config = new TelemetryConfiguration(string.Empty, new StubTelemetryChannel());
+                string testerKey = "tester123";
+                Assert.IsTrue(hbeat.AddHealthProperty(testerKey, "test", true));
+                hbeat.Initialize(configuration: config);
```

Comment:
do you need initialize() to be called here?

Reply:
Yes. Initialize also ensures the default SDK-provided payload gets added to the heartbeat. I wanted to ensure that Initialization doesn't clear user provided properties.